File: app/src/main/java/ru/bloknot/news/activity/FullActivity.java
package ru.bloknot.news.activity;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Intent;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Bundle;
import android.view.View;
import android.webkit.WebChromeClient;
import android.webkit.WebResourceRequest;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;

import java.io.IOException;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import ru.bloknot.news.R;

///* Открой другой проект app и тебе надо синхронизировать данные с этим классом.
/// А также надо найти способ передавать полные ссылки новостей

public class FullActivity extends Activity {

    WebView mWebView;
    private ProgressBar progressBar;
    TextView textView;
    LinearLayout layout;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_full);
        textView = findViewById(R.id.textLoad);
        layout = findViewById(R.id.linerNew);
        String url = getIntent().getStringExtra("url");
        System.out.println("Принял url: " + url);

        if (url != null && !url.isEmpty()) {
            setupWebView(url);
            System.out.println("Не нулл url продолжаем работу");
        }else {
            Toast.makeText(this, "Ошибка загрузки данных!", Toast.LENGTH_SHORT).show();
        }
    }

    @SuppressLint("SetJavaScriptEnabled")
    private void setupWebView(String url) {
        mWebView = findViewById(R.id.webView);
        progressBar = findViewById(R.id.progressBar);

        WebSettings settings = mWebView.getSettings();
        settings.setJavaScriptEnabled(false);
        settings.setLoadsImagesAutomatically(true);
        settings.setUserAgentString(getResources().getString(R.string.user_agent));
        mWebView.setWebChromeClient(new MyWebChromeClient());
        mWebView.setWebViewClient(new MyWebViewClient());
        downloadAndProcessHTML(url);
    }

    private class MyWebChromeClient extends WebChromeClient {
        public void onProgressChanged(WebView view, int progress) {

            String prog = "Загрузка " + progress + " %";
            textView.setText(prog);

            progressBar.setProgress(progress);

            if (progress == 100) {
                mWebView.setVisibility(View.VISIBLE);
                layout.setVisibility(View.GONE);
                progressBar.setVisibility(View.GONE);
            } else {
                layout.setVisibility(View.VISIBLE);
            }
        }
    }

    private class MyWebViewClient extends WebViewClient {
        @Override
        public boolean shouldOverrideUrlLoading(WebView view, WebResourceRequest request) {
            String url = request.getUrl().toString();

            if (url.contains("m.bloknot-krasnodar.ru") || url.contains("bloknot-krasnodar.ru")) {
                view.loadUrl(url);
                return true;
            }else {
                System.out.println("Ошибка!");
            }

            Intent intent = new Intent(Intent.ACTION_VIEW);
            intent.setData(Uri.parse(url));
            startActivity(intent);
            return true;
        }
        @SuppressLint("SetJavaScriptEnabled")
        @Override
        public void onPageStarted(WebView view, String url, Bitmap favicon) {
            view.getSettings().setJavaScriptEnabled(false);
            super.onPageStarted(view, url, favicon);
        }
    }

    /// Новый код

    private void downloadAndProcessHTML(String urlSite) {
        OkHttpClient client = new OkHttpClient();
        Request request = new Request.Builder()
                .addHeader("User-Agent", getResources().getString(R.string.user_agent))
                .url(urlSite)
                .build();

        client.newCall(request).enqueue(new Callback() {
            @Override
            public void onFailure(@NonNull Call call, @NonNull IOException e) {
                System.out.println("Ошибка: " + e.getMessage());
            }

            @Override
            public void onResponse(@NonNull Call call, @NonNull Response response) throws IOException {
                System.out.println("responce: " + response);

                if (response.code() == 404 || response.message().equals("Not Found")){
                    System.out.println("Ошибка 404");
                }

                if (!response.isSuccessful()) throw new IOException("Ошибка сервера");


                String htmlContent = response.body().string();
                modifyAndLoadHTML(htmlContent);
            }
        });
    }

    private void modifyAndLoadHTML(String htmlContent) {
        Document document = Jsoup.parse(htmlContent);

        /// Удаляем все ссылки
        document.select("a[href]").remove();

        // Удаляем меню верхнее полностью
        Element header = document.getElementById("header");

        if (header != null) {
            header.remove();
        }else {
            System.out.println("header: Не получилось удалить");
        }

        // Удаляем блоки опросов и другие ненужные секции
        Element poll = document.getElementById("pollFrame");

        if (poll != null) {
            poll.remove();
        }else {
            System.out.println("poll: Не получилось удалить");
        }

        document.getElementsByClass("flex.justify-between.bg-white.p-4.mb-2.border-b").remove();
        document.getElementsByTag("section").remove();
        document.getElementsByTag("h1").remove();
        document.getElementsByTag("footer").remove();
        document.getElementsByClass("text_below_content").remove();
        document.getElementsByTag("i").remove();
        document.getElementsByTag("iframe").remove();

        // Меняем стили
        document.head().append("<style> .pt-44{ padding-top:0.0rem; } </style>");
        Element main = document.select("main").first();
        if (main != null){
            main.removeAttr("style");
        }

///        System.out.println("----------------------");
///        System.out.println("main: " + main);
///       System.out.println("----------------------");

        final String modifiedHtml = document.html();

        runOnUiThread(() -> {
            mWebView.loadDataWithBaseURL("https://bloknot-krasnodar.ru/", modifiedHtml, "text/html", "utf-8", null);
        });
    }

    @Override
    public void onBackPressed() {
        finish();
        super.onBackPressed();
    }
}

File: app/src/main/java/ru/bloknot/news/activity/FullActivity.kt
// File: FullActivity.kt
package ru.bloknot.news.activity

import android.annotation.SuppressLint
import android.os.Bundle
import android.view.View
import android.webkit.WebChromeClient
import android.webkit.WebSettings
import android.webkit.WebView
import android.widget.ProgressBar
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import okhttp3.Call
import okhttp3.Callback
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import org.jsoup.Jsoup
import ru.bloknot.news.R
import ru.bloknot.news.utils.Constants
import java.io.IOException

class FullActivity : AppCompatActivity() {

    private lateinit var webView: WebView
    private lateinit var progressBar: ProgressBar
    private val client = OkHttpClient()

    @SuppressLint("SetJavaScriptEnabled")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_full)

        webView = findViewById(R.id.webView)
        progressBar = findViewById(R.id.progressBar)

        // Отключаем JS — нам он не нужен, только тормозит
        webView.settings.javaScriptEnabled = false
        webView.settings.cacheMode = WebSettings.LOAD_NO_CACHE

        val url = intent.getStringExtra("url") ?: run {
            Toast.makeText(this, "Ошибка: нет ссылки", Toast.LENGTH_SHORT).show()
            finish()
            return
        }

        val title = intent.getStringExtra("title") ?: run {
            Toast.makeText(this, "Ошибка: нет тайтла", Toast.LENGTH_SHORT).show()
            finish()
            return
        }

        webView.webChromeClient = object : WebChromeClient() {
            override fun onProgressChanged(view: WebView?, newProgress: Int) {
                super.onProgressChanged(view, newProgress)
                progressBar.visibility = View.VISIBLE

                progressBar.progress = newProgress // обновляем ProgressBar
                setTitle("Загрузка $newProgress %")

                if (newProgress == 100) {
                    setTitle(title)
                    progressBar.visibility = View.GONE      // скрываем, когда загрузка завершена
                    webView.visibility = View.VISIBLE
                } else {
                    progressBar.visibility = View.VISIBLE
                }
            }
        }

        loadMobileArticle(url)
    }

    private fun loadMobileArticle(url: String) {

        val request = Request.Builder()
            .url(url)
            // КЛЮЧ — правильный мобильный User-Agent (именно он заставляет сайт отдать m-версию)
            .header("User-Agent", Constants.USER_AGENT_MOBILE)
            .header("Accept", "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8")
            .header("Accept-Language", "ru-RU,ru;q=0.9")
            .build()

        client.newCall(request).enqueue(object : Callback {

            override fun onFailure(call: Call, e: IOException) {
                runOnUiThread {
                    progressBar.visibility = View.GONE
                    Toast.makeText(this@FullActivity, "Нет интернета", Toast.LENGTH_LONG).show()
                }
            }

            override fun onResponse(call: Call, response: Response) {
                response.use { resp ->
                    if (!resp.isSuccessful) {
                        runOnUiThread { Toast.makeText(this@FullActivity, "Ошибка ${resp.code}", Toast.LENGTH_SHORT).show() }
                        return
                    }

                    val html = resp.body.string()
                    val cleanHtml = cleanHtmlForMobile(html)

                    runOnUiThread {
                        webView.loadDataWithBaseURL(
                            Constants.BASE_URL,
                            cleanHtml,
                            "text/html",
                            "UTF-8",
                            null
                        )
                    }
                }
            }
        })
    }

    private fun cleanHtmlForMobile(html: String): String {
        if (html.isBlank()) return "<html><body>Пустая страница</body></html>"

        val doc = try {
            Jsoup.parse(html)
        } catch (e: Exception) {
            return "<html><body>Ошибка обработки страницы</body></html>"
        }



        // Объединяем удаление в один проход (быстрее и читаемее)
        doc.select("""
        script, noscript, iframe, 
        header, footer, nav, aside, 
        .advert, .banner, .reklama, 
        [style*=position:fixed], [class*=ad], [id*=ad], 
        i, p img[src*='plashka-novaya.jpg']""").remove()

        // Удаляем всё после рекламного <p> (включая сам <p>)
        doc.select("p:contains(Читайте новости Краснодара и края в удобном формате в нашем Телеграм-канале и в MAX.)")
            .firstOrNull()
            ?.let { adNode ->
                adNode.nextElementSiblings().forEach { it.remove() }
                adNode.remove()
            }

        // Отключаем все ссылки одним проходом (чисто и эффективно)
        doc.select("a[href]").forEach { link ->
            link.removeAttr("href")
                .removeAttr("target")
                .removeAttr("rel")
                .tagName("span")  // опционально — если хочешь полностью убрать <a>
        }

        // Выбираем контент — более надёжно
        val article = doc.selectFirst("article")
            ?: doc.selectFirst(".news-text, .news-detail, .content, .text, .bloknot-detail-text")
            ?: doc.body()

        // Стили выносим в константу (можно даже в отдельный файл)
        val articleStyle = """
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <style>
            body { font-family: 'Roboto', sans-serif; line-height: 1.6; padding: 16px; margin: 0; background: white; color: black; }
            
            img { 
                max-width: 100%; 
                height: auto; 
                border-radius: 12px; 
                margin: 20px 0; 
                box-shadow: 0 4px 12px rgba(0,0,0,0.1);
            }
            
            h1, h2 { color: #1a1a1a; font-size: 14px;}
            p { font-size: 17px; margin: 16px 0; }
            a { color: #0066cc; }
        </style>
    """.trimIndent()

        return buildString {
            append("<!DOCTYPE html><html><head><meta charset=\"utf-8\">")
            append(articleStyle)
            append("</head><body>")
            append(article.html())
            append("</body></html>")
        }
    }

    override fun onDestroy() {
        webView.destroy()
        super.onDestroy()
    }
}